manager/allocator/network/port/port.go

```diff
@@ -0,0 +1,367 @@
+package port
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/docker/swarmkit/api"
+)
+
+const (
+	// DynamicPortStart is the start of the dynamic port range from which node
+	// ports will be allocated when the user did not specify a port.
+	DynamicPortStart uint32 = 30000
+
+	// DynamicPortEnd is the end of the dynamic port range from which node
+	// ports will be allocated when the user did not specify a port.
+	DynamicPortEnd uint32 = 32767
+
+	// The start of master port range which will hold all the allocation state
+	// of ports allocated so far regardless of whether it was user defined or
+	// not.
+	masterPortStart uint32 = 1
+
+	// The end of master port range which will hold all the allocation state of
+	// ports allocated so far regardless of whether it was user defined or not.
+	masterPortEnd uint32 = 65535
+)
+
+// ErrPortInUse is an error type returned if one of the requested ports is
+// already in use by something else.
+type ErrPortInUse struct {
+	port
+}
+
+// Error returns a formatted message explaining what published port is in use
+func (e ErrPortInUse) Error() string {
+	return fmt.Sprintf("port %d/%s is already reserved", e.port.port, e.port.protocol)
+}
+
+// IsErrPortInUse returns true if the error is of type ErrPortInUse
+func IsErrPortInUse(e error) bool {
+	_, ok := e.(ErrPortInUse)
+	return ok
+}
+
+// PortAllocator is an allocator component that manages the state of which
+// ports (and which protocols) are in use in the the cluster.
+//
+// The PortAllocator totally owns the Endpoint.Ports field. Nothing should be
+// added to Endpoint.Ports outside of the port allocator. If this rule is
+// followed, the the port allocator doesn't have to check for cases like two
+// services having the same port allocated; we created all of the fields, and
+// we can know they are consistent.
+type PortAllocator struct {
+	// ports maps the ports in use. essentially, using map as a set
+	ports map[port]struct{}
+}
+
+// port is the minimal representation for the PortAllocator of a single port.
+// being composed of 2 numeric types, this type is comparable and can be used
+// as a map key
+type port struct {
+	// protocol represented by this port space
+	protocol api.PortConfig_Protocol
```

Comment:
You can save a little space by having a struct with a port vector for each protocol:

```
type PortAllocator struct {
  udp map[port]struct{}
  tcp map[port]struct{}
  sctp map[port]struct{}
}
```

Reply:
It's possible map overhead removes this benefit, however.

Like we discussed, this is _probably_ sparse. I did some experimentation with using `big.Int` as a bit set and that is pretty minimal.

Reply:
It's easier to reason in the code about ports-as-a-value than it is to reason different-protocols-different-spaces. 

The overhead of the extra data in each port object is minimal compared to the overhead of literally all the other dumb inefficient crap I'm doing, so even if we optimized here, it would be the tiniest win.